// mf-voice-advisor-web/src/lib/api.ts
import type { 
  StartConversationResponse, 
  NextQuestionResponse,
  ReportData,
  AnswerData
} from 'shared/types';

// Vite proxies /api to the backend
const API_BASE = '/api';

export async function startConversation(): Promise<StartConversationResponse> {
  const res = await fetch(`${API_BASE}/conversation/start`, { method: 'POST' });
  if (!res.ok) throw new Error('Failed to start conversation');
  return res.json();
}

export async function sendAnswer(sessionId: string, answerText: string): Promise<NextQuestionResponse> {
  const res = await fetch(`${API_BASE}/conversation/${sessionId}/answer`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ answerText })
  });
  if (!res.ok) throw new Error('Failed to send answer');
  return res.json();
}

export async function getHistory(sessionId: string): Promise<{ history: AnswerData[], status: string }> {
  const res = await fetch(`${API_BASE}/conversation/${sessionId}/history`);
  if (!res.ok) throw new Error('Failed to get history');
  return res.json();
}

export async function generateReport(sessionId: string): Promise<ReportData> {
  const res = await fetch(`${API_BASE}/report/${sessionId}/generate`, { method: 'POST' });
  if (!res.ok) throw new Error('Failed to generate report');
  return res.json();
}

export async function getReport(sessionId: string): Promise<ReportData> {
  const res = await fetch(`${API_BASE}/report/${sessionId}`);
  if (!res.ok) throw new Error('Failed to get report');
  return res.json();
}
